test/api.spec.ts

```diff
@@ -15,143 +15,256 @@
+			);
 		}
 	});
 
-	it('Should fail to create a release bundle using a non-existing release ID', async function () {
-		// amd64-supervisor https://dashboard.balena-cloud.com/apps/1667442/releases/3023927
-		const sdk = SDK.getSdk();
-		if (process.env.UPDATE_MOCKS === 'true') {
-			await bundle.create({
-				sdk,
-				releaseId: 3023927,
-			});
-		} else {
-			await expect(
-				bundle.create({
-					sdk,
-					releaseId: 3023927,
-				}),
-			).to.be.rejectedWith(Error, 'Release not found.');
+	it('should throw an error if the release does not include any images', function () {
+		try {
+			const release = JSON.parse(releaseData) as SDK.Release;
+			release.release_image = undefined;
+			bundle.$validateRelease(release);
+			throw new Error(
+				'Expected function to throw an error, but none was thrown',
+			);
+		} catch (err) {
+			expect(err).to.be.an.instanceof(Error);
+			expect(err.message).to.equal(
+				'Release bundles can only be created from releases with successfully built images.',
+			);
 		}
 	});
 });
 
-describe('Basic apply release usage', function () {
-	it('Should apply the release bundle to a fleet', async function () {
-		const mockFile = `${this.test?.title.replace(/ /g, '-').toLowerCase()}.json`;
-		const nocksDefinitionsFile = await fs.readFile(
-			path.resolve(__dirname, 'mocks', mockFile),
-			'utf8',
-		);
-		const expectedApplication = JSON.parse(nocksDefinitionsFile)[0].response
-			.d[0] as SDK.Application;
-		const sdk = SDK.getSdk();
-		const bundleStream = createReadStream('./test/fixtures/release-bundle.tar');
-		await bundle.apply({
-			sdk,
-			application: process.env.APPLICATION_ID
-				? Number(process.env.APPLICATION_ID)
-				: expectedApplication.id,
-			stream: bundleStream,
-		});
+describe('Basic apply release usage', async function () {
+	let releaseData: string;
+	before(async function () {
+		const releaseFile = path.resolve(__dirname, 'fixtures', 'release.json');
+		releaseData = await readFile(releaseFile, 'utf8');
 	});
 
-	it('Should fail in applying the release bundle to a fleet with existing successful release', async function () {
-		const mockFile = `${this.test?.title.replace(/ /g, '-').toLowerCase()}.json`;
-		const nocksDefinitionsFile = await fs.readFile(
-			path.resolve(__dirname, 'mocks', mockFile),
-			'utf8',
+	it('Should be able to read the manifest and normalize it', function () {
+		const release = JSON.parse(releaseData) as SDK.Release;
+		const normalizedManifest = bundle.$normalizeManifest(release);
+		expect(normalizedManifest.semver_major).to.equal(release.semver_major);
+		expect(normalizedManifest.semver_minor).to.equal(release.semver_minor);
+		expect(normalizedManifest.semver_patch).to.equal(release.semver_patch);
+		const expectedReleaseImages: Array<{
+			image: SDK.Image;
+			service: string;
+		}> = [];
+		for (const releaseImage of release.release_image!) {
+			if (!Array.isArray(releaseImage.image)) {
```

Comment:
This is better expressed as an expectation, ie. “expect(releaseImage.image).to.be.an(‘array’)”, otherwise it looks like “setup code”. There are a few other similar cases just below and further down this file, same comment applies.